library/std/src/sys/pal/uefi/helpers.rs

```diff
@@ -221,3 +227,185 @@ pub(crate) fn runtime_services() -> Option<NonNull<r_efi::efi::RuntimeServices>>
     let runtime_services = unsafe { (*system_table.as_ptr()).runtime_services };
     NonNull::new(runtime_services)
 }
+
+pub(crate) struct DevicePath(NonNull<r_efi::protocols::device_path::Protocol>);
+
+impl DevicePath {
+    pub(crate) fn from_text(p: &OsStr) -> io::Result<Self> {
+        fn inner(
+            p: &OsStr,
+            protocol: NonNull<r_efi::protocols::device_path_from_text::Protocol>,
+        ) -> io::Result<DevicePath> {
+            let path_vec = p.encode_wide().chain(Some(0)).collect::<Vec<u16>>();
```

Comment:
This will ignore internal NULL bytes in the string, resulting in truncation.

Reply:
I am not sure I understand. Can you provide me with an example?

Reply:
Exactly, yes. We should return an error in that case instead, [like on Windows](https://github.com/rust-lang/rust/blob/5ced3dad5752cc9423a19e9a5aafad6e12ed3063/library/std/src/sys/pal/windows/mod.rs#L177-L196).

Reply:
I am not doing manual loop unrolling like the windows version since I am not sure if it is necessary.

Reply:
don't worry. the loop unrolling was from quite some time ago. we should test to see if it even does anything anymore.